Add VK_VALVE_mutable_descriptor_type

HansKristian-Work opened the pull request.

Joshua-Ashton commented on chapters/descriptorsets.txt:
```diff
@@ -2999,6 +3008,13 @@ endif::VK_VERSION_1_2,VK_EXT_descriptor_indexing[]
     pname:dstSet must: not have been allocated with a layout that included
     immutable samplers for pname:dstBinding
+ifdef::VK_VALVE_mutable_descriptor_type[]
+  * If sname:VkDescriptorSetLayoutBinding for pname:dstSet at pname:dstBinding
```
s/If/If the/

HansKristian-Work replied:
Fixed.

Joshua-Ashton commented on chapters/descriptorsets.txt:
```diff
@@ -2100,6 +2108,20 @@ pname:maxUpdateAfterBindDescriptorsInAllPools, or if fragmentation of the
 endif::VK_VERSION_1_2,VK_EXT_descriptor_indexing[]
 
+ifdef::VK_VALVE_mutable_descriptor_type[]
+If a pname::pPoolSizes[i] descriptor type is ename:VK_DESCRIPTOR_TYPE_MUTABLE_VALVE,
```
s/If a pname::pPoolSizes[i] descriptor type is/If a pname:pPoolSizes[i]'s elink:VkDescriptorType is/

HansKristian-Work replied:
Fixed.

Joshua-Ashton commented on chapters/descriptorsets.txt:
```diff
@@ -2100,6 +2108,20 @@ pname:maxUpdateAfterBindDescriptorsInAllPools, or if fragmentation of the
+which kind of ename:VK_DESCRIPTOR_TYPE_MUTABLE_VALVE descriptors can: be allocated from this pool entry.
+If slink:VkMutableDescriptorTypeCreateInfoVALVE does not exist in the pname:pNext chain,
+the descriptor pool allocates enough memory to be able to allocate a ename:VK_DESCRIPTOR_TYPE_MUTABLE_VALVE
+descriptor with any supported sname:VkMutableTypeDescriptorListVALVE.
```
replace with something like ``any supported elink:VkDescriptorType as a mutable descriptor``?

HansKristian-Work replied:
Fixed.

HansKristian-Work commented:
Added VK_DESCRIPTOR_SET_LAYOUT_CREATE_HOST_POOL_BIT_VALVE. I feel this is needed, because otherwise it becomes very awkward to express descriptor limits. If we want to create very large host only descriptor sets, we are currently limited by non-updateafterbind limits. The descriptor set layout won't know it's going to be used with HOST_ONLY pools, so vkGetDescriptorSetLayoutSupport might not do what we expect, with the descriptor set having this information, we can be clear that the set itself is completely distinct from other descriptor sets, I also added wording to ban the use of vkCreatePipelineLayout from such a set.

BNieuwenhuizen commented on chapters/descriptorsets.txt:
```diff
@@ -450,11 +493,100 @@ ifdef::VK_VERSION_1_2,VK_EXT_descriptor_indexing[]
+  * If pname:flags contains ename:VK_DESCRIPTOR_SET_LAYOUT_CREATE_PUSH_DESCRIPTOR_BIT_KHR,
+    pname:flags must: not contain ename:VK_DESCRIPTOR_SET_LAYOUT_CREATE_HOST_ONLY_POOL_BIT_VALVE.
+endif::VK_KHR_push_descriptor[]
+  * If any binding has a pname:descriptorType of ename:VK_DESCRIPTOR_TYPE_MUTABLE_VALVE,
```
This should probably be in the `VK_KHR_push_descriptor` ifdef? Also a single statement would be enough to prevent both from being set at the same time, so you can probably remove one.

BNieuwenhuizen commented on chapters/descriptorsets.txt:
```diff
@@ -450,11 +493,100 @@ ifdef::VK_VERSION_1_2,VK_EXT_descriptor_indexing[]
+    sname:VkMutableTypeDescriptorListVALVE structures.
+
+If pname:descriptorTypeListCount is zero or if this structure is not included in the pname:pNext chain,
+the slink:VkMutableTypeDescriptorListVALVE for each element is considered to be zero or `NULL` for each member.
```
Is there an use for this? mutable descriptors with 0 allowed types sounds useless?

HansKristian-Work replied:
It's mostly to match wording for bindless flags. It must be non-zero for mutable descriptor type, and must be zero for non-mutable.

BNieuwenhuizen commented on chapters/descriptorsets.txt:
```diff
@@ -3222,6 +3468,23 @@ endif::VK_VERSION_1_2,VK_EXT_descriptor_indexing[]
+    is ename:VK_DESCRIPTOR_TYPE_MUTABLE_VALVE, and the new active descriptor type
+    is ename:VK_DESCRIPTOR_TYPE_MUTABLE_VALVE, the
+    pname:pMutableDescriptorTypeLists for pname:srcBinding and pname:dstBinding must:
+    match exactly.
```
FWIW for radv it is enough if the destination list contains the active type (though with a difference in size we may need to fall back to the memcpy per descriptor).

HansKristian-Work replied:
Yes, this restriction is there mostly for performance concerns, so drivers don't need a slower fallback path.

BNieuwenhuizen commented on chapters/features.txt:
```diff
@@ -3789,6 +3789,33 @@ include::{generated}/validity/structs/VkPhysicalDevice4444FormatsFeaturesEXT.txt
+  * [[features-mutableDescriptorType]] pname:mutableDescriptorType indicates that
+    the implementation must: support using the elink:VkDescriptorType of
+    ename:VK_DESCRIPTOR_TYPE_MUTABLE_VALVE with at least the following
+    descriptor types:
```
I think this is the risky bit wrt support on other vendors. I think on d3d12 you essentially can have 2 pools, one for buffers and one for images? (not sure where the texel buffers fall). At the same time I think some VU above means an implementation cannot partially support the extension if they cannot support any combination of these?

Might be workaround-able in software in any case so maybe not worth worrying too much about.

HansKristian-Work replied:
No, there's just one pool. The split is between SRV/UAV/CBV and samplers. That's why SAMPLER is not included in the mandatory list.

BNieuwenhuizen reviewed (commented):
Some more comments/nits but none should be implementation blocking for RADV. So I guess that can count as my stamp of approval
